streamingccl: add stream ingestion processor

pbardea opened the pull request:
This commit introduces a processor that will ingest a stream given a
list of partition addresses. This initial version is the minimum changes
required to ingest these KVs, and as a result some simplifications were
made.

KVs are buffered until a checkpoint event is received (for any
partition). The KVs are currently also only ever flushed upon receiving
a checkpoint event.

The flow will terminate when the event channels provided by the
streaming client close, which is not expected to happen when normally
interaction with the ingestion job.

Closes https://github.com/cockroachdb/cockroach/issues/57398.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58513)
<!-- Reviewable:end -->

adityamaru reviewed (approved):
> Reviewed 5 of 5 files at r1, 2 of 2 files at r2, 13 of 13 files at r3, 8 of 12 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58513)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/streamingccl/stream_ingestion_processor.go, line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58513#-MQSb14L-b-DNynMK8rQ:-MQSb14L-b-DNynMK8rR:b-3pnw1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/446d3945ab0af0fc3afb30ac9b8248af303fabcd/pkg/ccl/streamingccl/stream_ingestion_processor.go#L60)):*
> ```Go
> 	output  execinfra.RowReceiver
> 
> 	// TODO: This doesn't yet use a buffering adder since the current
> ```

again for my understanding, we can't use a buffering adder because we want specific ts for each KV and not just the one ts associated with the buffering adder? Do we see an implementation in which we will be able to use the BA?
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_test.go, line 94 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58513#-MQSMAyY4HgWHUl-6jsU:-MQSMAyY4HgWHUl-6jsV:b-4q7vsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/446d3945ab0af0fc3afb30ac9b8248af303fabcd/pkg/ccl/streamingccl/stream_ingestion_processor_test.go#L94)):*
> ```
> split and scatter
> ```
nit: StreamIngestionProcessor
___
*[pkg/kv/bulk/sst_batcher.go, line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58513#-MQSIcQDAwgP5PRA0dPT:-MQSIcQDAwgP5PRA0dPU:blsaosn) ([raw file](https://github.com/cockroachdb/cockroach/blob/446d3945ab0af0fc3afb30ac9b8248af303fabcd/pkg/kv/bulk/sst_batcher.go#L115)):*
> ```
> ingest duplicate keys
> ```
for my own understanding, what was the rationale behind this configuration? do we expect to see duplicate keys? 


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58513)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/streamingccl/stream_ingestion_processor.go, line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58513#-MQSb14L-b-DNynMK8rQ:-MQSbiH98MHWR5HwLYRV:bm4cdax) ([raw file](https://github.com/cockroachdb/cockroach/blob/446d3945ab0af0fc3afb30ac9b8248af303fabcd/pkg/ccl/streamingccl/stream_ingestion_processor.go#L60)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

again for my understanding, we can't use a buffering adder because we want specific ts for each KV and not just the one ts associated with the buffering adder? Do we see an implementation in which we will be able to use the BA?
</blockquote></details>

Yep that's correct. I do think that  we could use a BA-like object. Whether that be extending the BA or creating another similar `BulkAdder`. The `BulkAdder` `Add(ctx context.Context, key roachpb.Key, value []byte)` method makes this a bit annoying at the moment. In particular, I think it would be useful to give us an easy way to avoid this buffer from getting too large if we receive a lot of KV events without a checkpoint event.
___
*[pkg/kv/bulk/sst_batcher.go, line 115 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/58513#-MQSIcQDAwgP5PRA0dPT:-MQSbUPs7okz0oZnsVoc:bad1rkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/446d3945ab0af0fc3afb30ac9b8248af303fabcd/pkg/kv/bulk/sst_batcher.go#L115)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

> ```
> ingest duplicate keys
> ```
for my own understanding, what was the rationale behind this configuration? do we expect to see duplicate keys? 
</blockquote></details>

Yes I think we do expect them. A CDC stream provides "at least once" guarantees which means that we might seem some duplicate KV events which is why we expect to see duplicate keys.


<!-- Sent from Reviewable.io -->

adityamaru commented on pkg/sql/rowexec/processors.go:
```diff
@@ -365,6 +365,9 @@ var NewSplitAndScatterProcessor func(*execinfra.FlowCtx, int32, execinfrapb.Spli
 // NewRestoreDataProcessor is implemented in the non-free (CCL) codebase and then injected here via runtime initialization.
 var NewRestoreDataProcessor func(*execinfra.FlowCtx, int32, execinfrapb.RestoreDataSpec, *execinfrapb.PostProcessSpec, execinfra.RowSource, execinfra.RowReceiver) (execinfra.Processor, error)
 
+// NewStreamIngestionDataProcessor is implemented in the non-free (CCL) codebase and then injected here via runtime initialization.
```
Just noticed this hasn't been hooked up above in the `NewProcessor()` method.

pbardea replied:
Thanks for catching -- done.

pbardea committed:
streamclient: update resolved timestamp event to use hlc.Timestamp

Release note: None

pbardea committed:
streamingccl: add stream ingestion processor

This commit introduces a processor that will ingest a stream given a
list of partition addresses. This initial version is the minimum
changes
required to ingest these KVs, and as a result some simplificiations
were
made.

KVs are buffered until a checkpoint event is received (for any
partition). The KVs are currently also only ever flushed upon receiving
a checkpoint event.

The flow will terminate when the event channels provided by the
streaming client close, which is not expected to happen when normally
interactin with the ingestion job.

Release note: None

adityamaru commented:
LGTM 👍

pbardea commented:
TFTR!
bors r=adityamaru
